src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs

```diff
@@ -17,67 +17,78 @@
+                if (instrumentation is IDisposable disposable)
                 {
-                    disposableInstrumentation.Dispose();
+                    disposable.Dispose();
                 }
             }
         }
 
-        private void Detector_OnReady(object sender, AssemblyLoadDetector.LoadDetectorReadyEventArgs e)
+        public void Add(InstrumentationInitializer loader)
         {
-            string detectorName = sender.GetType().Name;
+            new OnAssemblyLoadInitializer(this, loader);
+        }
 
-            Logger.Debug("Detector '{0}' executed load", detectorName);
+        void ILifespanManager.Track(object instance)
+        {
+            _instrumentations.Add(instance);
+        }
 
-            try
-            {
-                var instrumentation = e.Builder();
+        private class OnAssemblyLoadInitializer
+        {
+            private static readonly ILogger Logger = OtelLogging.GetLogger();
+            private readonly InstrumentationInitializer _instrumentationInitializer;
+            private readonly LazyInstrumentationLoader _manager;
+            private HashSet<string> _requiredAssemblies;
+            private int _loadedCount;
 
-                // We just track the instances to keep them alive.
-                // During the shut down event IDisposable types are also disposed.
-                // This behavior repliactes SDK internal logic.
-                _instrumentations.Add(instrumentation);
-            }
-            catch (Exception ex)
+            public OnAssemblyLoadInitializer(LazyInstrumentationLoader manager, InstrumentationInitializer instrumentationInitializer)
             {
-                Logger.Error(ex, "Could not execute '{0}' load", detectorName);
+                _instrumentationInitializer = instrumentationInitializer;
+                _manager = manager;
+                _requiredAssemblies = new HashSet<string>(instrumentationInitializer.RequiredAssemblies);
+                AppDomain.CurrentDomain.AssemblyLoad += CurrentDomain_AssemblyLoad;
             }
 
-            var detector = (AssemblyLoadDetector)sender;
-            detector.OnReady -= Detector_OnReady;
+            private void CurrentDomain_AssemblyLoad(object sender, AssemblyLoadEventArgs args)
+            {
+                string assemblyName = args.LoadedAssembly.FullName.Split(new[] { ',' }, count: 2)[0];
 
-            _assemblyLoadDetectors.Remove(detector);
+                if (_requiredAssemblies.Contains(assemblyName))
+                {
+                    if (Interlocked.Increment(ref _loadedCount) == _requiredAssemblies.Count)
+                    {
+                        string initializerName = _instrumentationInitializer.GetType().Name;
+                        Logger.Debug("'{0}' started", initializerName);
+
+                        try
+                        {
+                            _instrumentationInitializer.Initialize(_manager);
```

Comment:
I think we should make it async and enable parallel processing? as the `CurrentDomain_AssemblyLoad` seems to be synchronized.